flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/snapshot/RocksDBSnapshotStrategyBase.java

```diff
@@ -395,18 +394,16 @@ public void release() {
     /** Previous snapshot with uploaded sst files. */
     protected static class PreviousSnapshot {
 
-        @Nullable private final Map<StateHandleID, Long> confirmedSstFiles;
+        @Nullable private final Map<StateHandleID, StreamStateHandle> confirmedSstFiles;
 
-        protected PreviousSnapshot(@Nullable Map<StateHandleID, Long> confirmedSstFiles) {
+        protected PreviousSnapshot(
+                @Nullable Map<StateHandleID, StreamStateHandle> confirmedSstFiles) {
             this.confirmedSstFiles = confirmedSstFiles;
         }
 
         protected Optional<StreamStateHandle> getUploaded(StateHandleID stateHandleID) {
             if (confirmedSstFiles != null && confirmedSstFiles.containsKey(stateHandleID)) {
-                // we introduce a placeholder state handle, that is replaced with the
-                // original from the shared state registry (created from a previous checkpoint)
-                return Optional.of(
-                        new PlaceholderStreamStateHandle(confirmedSstFiles.get(stateHandleID)));
```

Comment:
Could you elaborate why you had to delete `PlaceholderStreamStateHandle`?

I'm concerned that with this change, `ByteStreamStateHandle` will always be sent always to the JM (regardless of whether they were previously sent or not).

Reply:
(otherwise, the branching code here can be simplified with `Optional.ofNullable`)

Reply:
ByteStreamStateHandle tends to occupy little memory (no more than 20KB by default), and these memory occupations only occur during network transmission. When repeated ByteStreamStateHandle is passed to SharedStateRegistry, it will be discarded. And, I think we should try to avoid the generation of small files (maybe through file aggregation, etc.).
Using PlaceholderStreamStateHandle makes it difficult to implement SharedStateRegisterKey based on physical id, although it can be achieved by storing physical id of original StateHandle into PlaceholderStreamStateHandle. But I still think removing the PlaceholderStreamStateHandle would make the SharedStateRegistry much simpler and clearer. Since I haven't observed bottlenecks in ByteStreamStateHandle delivery in job with 2.6TB of state, I think the rpc overhead of removing PlaceholderStreamStateHandle is acceptable.
Is the above explanation acceptable to you @rkhachatryan  ?

Reply:
Thanks for the explanation @zoltar9264 

My concern is exactly about additional overhead of data transmission (the purpose of placeholder is to reduce it).
I'm not sure whether `PlaceholderStreamStateHandle` removal will not cause a regression.
It depends not only on the state size, but also on the configuration (`state.storage.fs.memory-threshold`) and the actual size of the SST files.

So I think it would be safer to keep it. I agree that that requires more code changes, but I think it worth it.
The complexity added by `PlaceholderStreamStateHandle` doesn't seem very high.

WDYT?

Reply:
I really don't have a full test to prove that removing PlaceholderStreamStateHandle has no regression, I therefore agree to keep it. But I still recommend limiting its usage to ByteStreamStateHandle, and don't use it to calculate checkpointed size.
And I suggest saving the PhysicalStateHandleID of the original StateHandle into the PlaceholderStreamStateHandle to simplify the registration. And we can check it before replacing.
WDYT @rkhachatryan ?

Reply:
> But I still recommend limiting its usage to ByteStreamStateHandle, and don't use it to calculate checkpointed size.

IIUC, that's already implemented. Why would you change this?

> And I suggest saving the PhysicalStateHandleID of the original StateHandle into the PlaceholderStreamStateHandle to simplify the registration. 

That might not be necessary, if the registration key is the same key as in `IncrementalRemoteKeyedStateHandle.sharedState`, as per [comment](https://github.com/apache/flink/pull/22669#discussion_r1210450731).

Reply:
Thanks for your patience @rkhachatryan .

> But I still recommend limiting its usage to ByteStreamStateHandle, and don't use it to calculate checkpointed size.

Before this pr, RocksDBSnapshotStrategyBase use PlaceholderSteamStateHandle no matter the origin handle is ByteStreamStateHandle or FileStateHandle. I suggest only use PlaceholderStreamStateHandle while the origin handle is ByteStreamStateHandle. This pr already implemented not use PlaceholderStreamStateHandle calculate checkpointed size, I want keep it.

> That might not be necessary, if the registration key is the same key as in IncrementalRemoteKeyedStateHandle.sharedState, as per https://github.com/apache/flink/pull/22669#discussion_r1210450731.

Do you mean change IncrementalRemoteKeyedStateHandle like this ?
```
    /** Shared state in the incremental checkpoint. */
    private final Map<SharedStateRegistryKey, HandleWithLocalPath> sharedState;

    /** Private state in the incremental checkpoint. */
    private final List<HandleWithLocalPath> privateState;

    private static final class HandleWithLocalPath {
        StreamStateHandle handle;
        String localPath;
    }
```

Reply:
> Do you mean change IncrementalRemoteKeyedStateHandle like this ?

Yes, I mean something like that, maybe using `Path` or `File` instead of `String` for `localPath`.

> I suggest only use PlaceholderStreamStateHandle while the origin handle is ByteStreamStateHandle. This pr already implemented not use PlaceholderStreamStateHandle calculate checkpointed size, I want keep it.

Can you share the motivation?
I think that this will just add an additional `instanceof` and increase complexity. It would also easier to break if we add a new type of handle that needs replacement. Or am I missing something?

Reply:
@zoltar9264 I think localPath has already recorded in [StateHandleID](https://github.com/apache/flink/pull/22669/files#diff-b63b70b359ed2867033aac1ff62164658446483bf2d7a49cf18c35b6f448f089R401-R402), do we need to record an additional one? If `IncrementalRemoteKeyedStateHandle#sharedState` is modified, RocksDBDownloader also needs to modify accordingly.

Reply:
Hi @fredia , in the previous discussion, we mentioned that the registry key based on the remote file name (physical id) is used as the key of privateState.

Reply:
Thanks for reply @rkhachatryan. Which handles to use in checkpoint should be determined before registration. We use placeholder only to reduce network transmission overhead. When we don't use placeholder, if the handle returned by the registerReference method is not equal to the registered handle, it means that there may be a potential bug. For example, the registry in this issue is wrongly de-duplicated. If we only perform the replacement when the placeholder is used, and reduce the scope of the placeholder, it is possible to avoid the problem caused by the registry returning the wrong handle. Furthermore, if we check that the handle returned by the registry is equal to the registered handle when the placeholder is not used, we can detect the problem earlier.

Reply:
Thanks @zoltar9264 ,

> For example, the registry in this issue is wrongly de-duplicated.
> If we only perform the replacement when the placeholder is used, and reduce the scope of the placeholder, it is possible to avoid the problem caused by the registry returning the wrong handle.

But this issue is not related to the `PlaceholderStreamStateHandle`, is it? It happens when both new and old handles are NOT `PlaceholderStreamStateHandle`s.

> Furthermore, if we check that the handle returned by the registry is equal to the registered handle when the placeholder is not used, we can detect the problem earlier.

I think the problem is actually when the handles under the same key are NOT equal - and we need to resolve this conflict somehow. Note that's not related `PlaceholderStreamStateHandle`.

But since in this PR we entirely eliminate the possibility of collision by using unique keys, such a collision would mean a bug; so we can remove that resolution logic and raise an error instead:
- remove `SharedStateRegistryImpl.SharedStateEntry.confirmed` field
- remove any calls to `scheduleAsyncDelete()` from `registerReference()`
-  (but keep `PlaceholderStreamStateHandle`)

WDYT?

Reply:
Thanks for your patience in explaining @rkhachatryan , I totally agree with your last proposal.

And I realized in coding that maybe we shouldn't store SharedStateRegistryKey in IncrementalRemoteKeyedStateHandle. Because SharedStateRegistryKey is only used during registration and can be calculated from physical id. More importantly, when the remote state file is relocated (savepoint is relocatable), the physical id will change accordingly. The [aforementioned change of IncrementalRemoteKeyedStateHandle](https://github.com/apache/flink/pull/22669#discussion_r1213882810) should look like this:
```
    /** Shared state in the incremental checkpoint. */
    private final List<HandleAndLocalPath> sharedState;

    /** Private state in the incremental checkpoint. */
    private final List<HandleAndLocalPath> privateState;

    private static final class HandleAndLocalPath {
        StreamStateHandle handle;
        String localPath;
    }
```
This way, we don't need to introduce a new serialized version for IncrementalRemoteKeyedStateHandle in `MetadataV2V3SerializerBase` either.

Reply:
Thanks for yours too @zoltar9264 😂

Sounds good, I like this approach 👍